### PR TITLE
Migrations for all serializable objects

### PR DESCRIPTION
**Pull request by nick-potts**
This PR introduces a flexible data migration system for all serializable objects, ensuring backward compatibility and data integrity as structures evolve. 

## Key Features

- Works on States, Events, DTOs - basically any ```ShouldMigrateData``` class
- Define migrations as an array or a ```Migrations``` class
- Automatic version tracking
- Migrations run only once per object

## Usage

1. Implement ```ShouldMigrateData``` on target class
2. Place ```MigrationNormalizer``` above ```SelfSerializingNormalizer``` in ```verbs.php``` config
3. Define migrations (array or class)
4. Migrations auto-apply during deserialization of events, states and dtos

## Usage Examples

```php
class UserSignedUp extends Event implements ShouldMigrateData
{
    public static function migrations(): array|Migrations
    {
        return [
            0 => fn($data) => array_merge($data, ['username' => 'user_' . rand(1000, 9999)]));
        ];
    }
}

// Or using a Migrations class:
class UserSignedUpMigrations extends Migrations {
   // short names
   public function v0(array $data): array
    {
        return array_merge($data, ['initial_migration' => true]);
    }
   // Snake Case
    public function v1_add_username(array $data): array
    {
        return array_merge($data, ['username' => 'user_' . rand(1000, 9999)]);
    }
    // Camel Case
    public function v2UpdateEmailFormat(array $data): array
    {
        $data['email'] = strtolower($data['email']);
        return $data;
    }
}
```

It currently works on all objects that implement the ShouldMigrateData interface. So it'll migrate a DTO in the constructor of an event, or a DTO in a state.

## Tidbits
Just a few random thoughts:
- Consider starting migration numbers from 1 instead of 0 to prevent potential issues with array indexing
- The version key is currently named `__vn` - this naming could be revisited
- The system currently runs all migrations beyond the max version without validating previous migrations
- Test coverage is mostly complete but could benefit from reorganization
- Explore the possibility of a trait for Events to skip migrations on deserialization and apply them during Event replay instead, that way you could inject States into migrations
- Since this requires a new normalizer, should the `SelfSerializingNormalizer` check for the __vn key and throw an error to warn about the MigrationNormalizer not being found?

### CODE REVIEW

**Comment by nick-potts** [comment]
Trying a more 'Verbsy' way of doing migrations. 

Added attributes for PropertyAdded, PropertyRemoved, PropertyMigrated. Since migrations are a process that occurs over time, you can simply generate migrations for a property when you need it.

Need a new property?

```php
#[PropertyAdded(version: 1, value: 'email@example.com')]
public string $email
```

Let's say we have a UserDto class with $email and $full_name properties:

```php
class UserDto
{
    public string $email;
    public string $full_name;
}
```

Now, we want to split the $full_name into $first_name and $last_name, and remove the original $full_name property:

```php
#[PropertyRemoved(version: 2, property: 'full_name')]
class UserDto implements ShouldMigrateData
{
    use HasMigrations;

    public string $email;
    public string $first_name;
    public string $last_name;

    #[PropertyAddedUsing(version: 0, property: 'first_name')]
    public function addFirstName(array $data)
    {
        return explode(' ', $data['full_name'], 2)[0];
    }

    #[PropertyAddedUsing(version: 1, property: 'last_name')]
    public function addLastName(array $data)
    {
        return explode(' ', $data['full_name'], 2)[1] ?? '';
    }
}
```

So in total you'd have 3 different ways to do migrations - via an array of closures, a dedicated migrations class or directly on properties with attributes.

There might be a way to inject state into those methods, but that is beyond my current understanding of Verbs.

File: src/Attributes/Migrations/MigrateUsing.php [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+        $reflect = new \ReflectionClass($target);
+
+        $method = $reflect->getMethod($using);
+
+        $instance = $reflect->newInstanceWithoutConstructor();
+        $method = $method->getClosure($instance);
+        try {
+            $data[$this->property] = $method($data);
+        } catch (\TypeError $e) {
+            throw new MigratorException("Method $using should accept an array and return the new value for $this->property.");
```
Comment (nick-potts) [inline src/Attributes/Migrations/MigrateUsing.php:26]: I'm in two minds over whether a `using` attribute should return just the value, or all the data.

**Comment by inxilpro** [comment]
This is an interesting approach. I'm going to have to mull it over for a bit, but it's cool!

**Comment by nick-potts** [comment]
I'm not sold on it either, I think in the ideal world migrations would get run before the apply with access to stuff from the container.

**Comment by DanielCoulbourne** [comment]
@nick-potts just got off a call with @inxilpro 

For now we want to keep the API really simple, how does something like this feel?

```php
   public static function migrate()
    {
        return [
            1 => fn(Collection $v0) => $v0->except('user_id'),
            2 => fn(Collection $v1) => $v1->merge(['email' => 'default@email.gov']),
        ];
    }
```

**Comment by nick-potts** [comment]
Probably best keeping it simple, it's quite clear how it works off at first glance and would just be another step in the lifecycle - I like.

**Pull request closed without merging**